core/src/main/java/org/apache/iceberg/util/Tasks.java

```diff
@@ -570,7 +566,7 @@ public Integer next() {
   }
 
   public static <I> Builder<I> foreach(Stream<I> items) {
-    return new Builder<>(items::iterator);
+    return new Builder<>(items.collect(Collectors.toList()));
```

Comment:
Our "iterator<I> items" already makes an iterable out of an iterator so I don't think we should avoid it here. This change also forces a materialization so we should avoid it.

Reply:
as ArrowReader，revert this.